### ISSUE TITLE
kernel_product outputs NaN's values when using a kernel raised to an integer power

### ISSUE THREAD

**Issue by little-nem** [getkeops/keops#112]
Hi KeOps team!
I am facing an issue involving the `kernel_product` function, which tends to return tensors with `nan`'s elements when using kernels that have been raised to an integer exponent. I managed to narrow down the issue to the following short code which demonstrates it.

```python
import torch
from pykeops.torch import Kernel, kernel_product

params_linear_kernel = {
    "id"      : Kernel("linear(x,y)"),
    "gamma"   : ( None ),
}

params_squared_kernel = {
    "id"      : Kernel("linear(x,y)**2"),
    "gamma"   : ( None ),
}

x = torch.tensor([[-1.],[1.]])
y = torch.tensor([[1.], [1.]])
b = torch.tensor([[1.], [1.]])

result_linear_negative_positive = kernel_product(params_linear_kernel, x, y, b)
result_squared_negative_positive = kernel_product(params_squared_kernel, x, y, b)
result_squared_positive_positive = kernel_product(params_squared_kernel, y, y, b)

result_linear_negative_positive, result_squared_negative_positive, result_squared_positive_positive
```
Which output is the following:
```python
(tensor([[-2.],
         [ 2.]]),
 tensor([[nan],
         [2.]]),
 tensor([[2.],
         [2.]]))
```
The first entry of the `result_squared_negative_positive`, which should have been equal to `(-1 × 1)**2 + (-1 × 1)**2`, ends up being equal to `nan`. Seeing that the other examples were computed just fine, I guess this should point to an issue in the implementation of exponentiation for negative entries.

Trying to make sure that I was not doing anything “unintended” with my code, I also ran the last example found [here](https://www.kernel-operations.io/keops/python/kernel-product.html), which also uses a squared linear kernel and exhibits the same behaviour: the final tensor `a` is full of `nan`'s.

If that is relevant: I am using `pykeops v1.4.1` available via `pip`.

Thank you!
Nemo

**Comment by jeanfeydy** [comment]
Hi @little-nem ,

Thanks for your report!
You were right: the problem came from CUDA's [powf](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__SINGLE.html#group__CUDA__MATH__SINGLE_1gb519b517c0036b3604d602f716a919dd) function, whose behaviour is not super well documented for integer inputs. It is both slower than a simple "squaring" operation, and returns NaNs if the exponent is not cast from int to float properly. As a consequence, **I added a special case for the squaring operation** in the `kernel_product` helper to fix the issue.

Note, however, that the `kernel_product` API has been made **obsolete** by the `LazyTensor` wrapper, and will be deprecated in future releases. Please consider switching to a LazyTensor implementation using e.g. the `(x_i | y_j) ** 2` syntax : your code will be much clearer and easy to maintain in the long run :-)

Best regards, and see you soon,
Jean

**Comment by little-nem** [comment]
Hi  @jeanfeydy ,
Thank you for your answer, and such a quick fix! :-)
Anyway, as you suggested, I took the time to switch to a full `LazyTensor`-based code which is indeed clearer! I do not know why I chose to go with the `kernel_product` API for most of my code in the first place; maybe it could be worth just mentioning somewhere in the corresponding documentation that this API is intended to be deprecated to avoid other people going the same way as I did.

Best, and see you soon,
Nemo
